tutorials/detailed_model_reference.ipynb

```diff
@@ -0,0 +1,663 @@
+    "    Args:\n",
+    "        rel_sampling_ratios (tuple): relative ratios for successive farthest point sampling\n",
+    "        interp_simplex (str): reference simplex for barycentric interpolation ('triangle' or 'tetrahedron')\n",
+    "    \"\"\"\n",
+    "\n",
+    "    def __init__(self, rel_sampling_ratios: tuple, interp_simplex: str):\n",
+    "        self.rel_sampling_ratios = rel_sampling_ratios\n",
+    "        self.interp_simplex = interp_simplex\n",
+    "\n",
+    "        self.dim_interp_simplex = {'triangle': 2, 'tetrahedron': 3}[interp_simplex]\n",
+    "\n",
+    "    def __call__(self, data: pyg.data.Data) -> Data:\n",
+    "\n",
+    "        pos = data.pos\n",
+    "        batch = data.surface_id.long() if hasattr(data, 'surface_id') else torch.zeros(pos.size(0), dtype=torch.long)\n",
+    "\n",
+    "        for i, sampling_ratio in enumerate(self.rel_sampling_ratios):\n",
+    "\n",
+    "            sampling_idcs = fps(pos, batch, ratio=sampling_ratio)  # takes some time but is worth it\n",
+    "            # sampling_idcs = torch.arange(0, pos.size(0), 1. / sampling_ratio, dtype=torch.int)\n",
+    "\n",
+    "            pool_source, pool_target = knn(pos[sampling_idcs], pos, 1, batch[sampling_idcs], batch)\n",
+    "            interp_target, interp_source = knn(pos[sampling_idcs], pos, self.dim_interp_simplex + 1, batch[sampling_idcs], batch)\n",
+    "\n",
+    "            data[f'scale{i}_pool_target'], data[f'scale{i}_pool_source'] = pool_target.int(), pool_source.int()\n",
+    "            data[f'scale{i}_interp_target'], data[f'scale{i}_interp_source'] = interp_target.int(), interp_source.int()\n",
+    "            data[f'scale{i}_sampling_index'] = sampling_idcs.int()\n",
+    "\n",
+    "            pos = pos[sampling_idcs]\n",
+    "            batch = batch[sampling_idcs]\n",
+    "\n",
+    "        return Data(**data)\n",
+    "\n",
+    "    def __repr__(self) -> str:\n",
+    "        return f\"{self.__class__.__name__}(rel_sampling_ratios={self.rel_sampling_ratios}, interp_simplex={self.interp_simplex})\""
+   ],
+   "outputs": [],
+   "execution_count": 3
+  },
+  {
+   "metadata": {
+    "ExecuteTime": {
+     "end_time": "2024-07-12T12:16:39.630319Z",
+     "start_time": "2024-07-12T12:16:39.616800Z"
+    }
+   },
+   "cell_type": "code",
+   "source": [
+    "transform = PointCloudPoolingScales(rel_sampling_ratios=(0.2,), interp_simplex='triangle')\n",
+    "dummy_data = transform(pyg.data.Data(pos=pos, orientation=orientation, scalar_feature=scalar_feature))"
+   ],
+   "id": "54400f8d7cb85dd8",
+   "outputs": [],
+   "execution_count": 4
+  },
+  {
+   "metadata": {},
+   "cell_type": "markdown",
+   "source": "## 2. Building the Model",
+   "id": "15893f75d230b52e"
+  },
+  {
+   "cell_type": "markdown",
+   "id": "546e0c748430850a",
+   "metadata": {},
+   "source": [
+    "A geometric algebra interface to embed your data in $\\mathbf{G}(3, 0, 1)$.\n",
+    "\n",
+    "```TODO```"
```

Comment:
Simply giving input and output channels is not enough when working with geometric algebra. Instead, we must define how our data is going to be represented by multivectors (and auxiliary scalars). For convenience, we package this in a `GeometricAlgebraInterface`. Here, we embed the mesh surface normal as planes at each vertex and pass an addtional scalar feature.

At the model output, we "translate" the multivectors back into Euclidean vectors by interpreting them as translation (pun intended). There are multiple modelling choices for how to perform this extraction (see [paper](https://arxiv.org/abs/2403.07536)).